File: Back Tracking/排列/47.全排列-ii.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <math.h>

using namespace std;
/*
 * @lc app=leetcode.cn id=47 lang=cpp
 *
 * [47] 全排列 II
 */

// @lc code=start
class Solution
{
public:
    vector<vector<int>> result; // 存放结果
    vector<int> path;
    void backtracking(vector<int> &nums, vector<bool> used)
    {
        if (path.size() == nums.size())
        {
            result.push_back(path);
            return;
        }

        for (int i = 0; i < nums.size(); i++)
        {
            if (used[i])
            {
                continue;
            }

            // nums[i] == nums[i - 1] && used[i - 1] == false 表示同一层有重复的元素
            if (i > 0 && nums[i] == nums[i - 1] && used[i - 1] == false)
            {
                continue;
            }
            path.push_back(nums[i]);
            used[i] = true;
            backtracking(nums, used);
            used[i] = false;
            path.pop_back();
        }
    }
    vector<vector<int>> permuteUnique(vector<int> &nums)
    {
        vector<bool> used(nums.size(), false);

        // 首先把给nums排序，让其相同的元素都挨在一起。
        sort(nums.begin(), nums.end());
        backtracking(nums, used);

        return result;
    }
};
// @lc code=end

File: Back Tracking/排列/47.全排列-ii.ts
/*
 * @lc app=leetcode.cn id=47 lang=typescript
 *
 * [47] 全排列 II
 */

// @lc code=start
function permuteUnique(nums: number[]): number[][] {
  const result: any = []
  const length = nums.length
  const used = {}
  if (length === 0) {
    return result
  }

  const backTracking = (arr) => {
    if (arr.length === length) {
      result.push([...arr])
      return
    }

    for (let i = 0; i < length; i++) {
      // 数组中不存在重复元素，所以如果已经被使用了，不在放入arr
      if (used[i]) {
        continue
      }
      // 如果相邻的两个元素相同，同时上一个元素跟当前元素不在同一次DFS中，则跳过
      // 例如211，遍历第一个1时候下一次回溯已经会包含第二个1，第2个1直接跳过，避免重复
      if (i > 0 && nums[i] == nums[i - 1] && used[i - 1] === false) {
        continue
      }
      arr.push(nums[i])
      used[i] = true
      backTracking(arr)
      used[i] = false
      arr.pop()
    }
  }
  // 先对数据排序，方便去重
  nums.sort((a, b) => a - b)
  backTracking([])

  return result
};
// @lc code=end

